pkg/platform/kube/scaler/scaler.go

```diff
@@ -0,0 +1,142 @@
+	"github.com/nuclio/nuclio/pkg/platformconfig"
+	"github.com/nuclio/nuclio/pkg/version"
+
+	"github.com/nuclio/logger"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	metricsv1 "k8s.io/metrics/pkg/client/clientset_generated/clientset"
+	custommetricsv1 "k8s.io/metrics/pkg/client/custom_metrics"
+)
+
+type metricEntry struct {
+	timestamp    time.Time
+	value        int64
+	functionName string
+	metricName   string
+}
+
+type Scaler struct {
+	logger                 logger.Logger
+	namespace              string
+	restConfig             *rest.Config
+	kubeClientSet          kubernetes.Interface
+	metricsPoller          *metricsPoller
+	metricsClientset       *metricsv1.Clientset
+	customMetricsClientSet custommetricsv1.CustomMetricsClient
+	nuclioClientSet        *nuclioio_client.Clientset
+	autoscaler             *Autoscale
+}
+
+func NewScaler(parentLogger logger.Logger,
+	namespace string,
+	kubeClientSet kubernetes.Interface,
+	metricsClientSet *metricsv1.Clientset,
+	nuclioClientSet *nuclioio_client.Clientset,
+	customMetricsClientSet custommetricsv1.CustomMetricsClient,
+	platformConfig *platformconfig.Configuration) (*Scaler, error) {
+
+	// replace "*" with "", which is actually "all" in kube-speak
+	if namespace == "*" {
+		namespace = ""
+	}
+
+	scaler := &Scaler{
+		logger:                 parentLogger,
+		namespace:              namespace,
+		kubeClientSet:          kubeClientSet,
+		metricsClientset:       metricsClientSet,
+		customMetricsClientSet: customMetricsClientSet,
+		nuclioClientSet:        nuclioClientSet,
+	}
+
+	var err error
+
+	// a shared metricsChannel between autoscaler and sources
+	metricsChannel := make(chan metricEntry)
+
+	metricName := platformConfig.ScaleToZero.MetricName
+	scaleWindow, err := time.ParseDuration(platformConfig.ScaleToZero.WindowSize)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to parse scale window size")
+	}
+
+	metricInterval, err := time.ParseDuration(platformConfig.ScaleToZero.PollerInterval)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to parse metric poller interval")
+	}
+
+	scaleInterval, err := time.ParseDuration(platformConfig.ScaleToZero.ScalerInterval)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to parse scaler interval")
+	}
+
+	err = scaler.createAutoScaler(metricsChannel, metricName, scaleWindow, scaleInterval)
+	if err != nil {
+		return nil, err
+	}
+
+	scaler.metricsPoller, err = NewMetricsPoller(scaler.logger, scaler, metricsChannel, metricInterval, metricName, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// log version info
+	version.Log(scaler.logger)
+	return scaler, nil
+}
+
+func (s *Scaler) createAutoScaler(metricsChannel chan metricEntry,
+	metricName string,
+	scaleWindow time.Duration,
+	scaleInterval time.Duration) error {
+	scalerFunction := func(namespace string, functionName string) {
```

Comment:
use interfaces, not callbacks. define `functionScaler` interface that has `scaleFunctionToZero` and add this to `Scaler`.